### PR TITLE
SJQ2T 428: BE Revoke Files Permissions

### PR DESCRIPTION
**Pull request by DarkoVeljanoskiHtec**

### CODE REVIEW

File: backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java [hunk 1]
```diff
@@ -237,6 +237,21 @@ public void revokePermission(UUID bucketId, UUID userId) {
...
   }
 
+  @Override
+  public void revokeFilePermission(UUID fileId, UUID userId) throws FileNotFoundException {
+    FileMetadataEntity fileMetadataEntity =
+        fileRepository.findById(fileId).orElseThrow(FileNotFoundException::new);
+    checkIfBucketIsDeleted(fileMetadataEntity.getBucket().getId());
+    currentUserHasPermissionForFile(fileMetadataEntity, AccessType.OWNER);
+    currentUserHasPermissionForFolder(fileMetadataEntity.getParentFolder(), AccessType.OWNER);
+    currentUserHasPermissionForBucket(fileMetadataEntity.getBucket().getId(), AccessType.OWNER);
```
Comment (ivanfhtec) [inline backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java]: I don't think these last two checks are necessary
`currentUserHasPermissionForFile()` should be enough

File: backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java [hunk 2]
```diff
@@ -237,6 +237,21 @@ public void revokePermission(UUID bucketId, UUID userId) {
...
+
+    UserEntity user = userRepository.findById(userId).orElseThrow(UserNotFoundException::new);
+    UserObjectAccessEntity permission =
+        permissionRepository.findByObjectIdAndGrantedTo(fileId, user);
```
Comment (ivanfhtec) [inline backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java]: this doesn't return an optional? should we refactor?
if this returns a null, we should throw an error?

Comment (DarkoVeljanoskiHtec) [reply backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java]: It does return Optional, but it was asking  for throw in method signature as well, I can't say why.

File: backend/src/main/java/com/htecgroup/skynest/service/PermissionService.java [hunk 3]
```diff
@@ -29,6 +30,8 @@ public interface PermissionService {
 
   void revokePermission(UUID bucketId, String userEmail);
 
+  void revokeFilePermission(UUID fileId, String email) throws FileNotFoundException;
```
Comment (MarkoRadunovicHtec) [inline backend/src/main/java/com/htecgroup/skynest/service/PermissionService.java]: Why is throws declaration there?

Comment (DarkoVeljanoskiHtec) [reply backend/src/main/java/com/htecgroup/skynest/service/PermissionService.java]: Intelij is asking for it, I don't really understand why.

File: backend/src/main/java/com/htecgroup/skynest/service/impl/BucketServiceImpl.java [hunk 4]
```diff
@@ -124,7 +124,9 @@ public List<BucketResponse> listAllBuckets() {
 
     List<BucketEntity> entityList =
         (List<BucketEntity>) bucketRepository.findAllByOrderByNameAscCreatedOnDesc();
-    entityList.stream().filter(this::doesCurrentUserHaveViewPermissionOnBucket);
+    entityList.stream()
+        .filter(this::doesCurrentUserHaveViewPermissionOnBucket)
+        .collect(Collectors.toList());
```
Comment (ivanfhtec) [inline backend/src/main/java/com/htecgroup/skynest/service/impl/BucketServiceImpl.java:129]: shouldn't this be
```
entityList = entityList.stream()
        .filter(this::doesCurrentUserHaveViewPermissionOnBucket)
        .collect(Collectors.toList());
```

File: backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java [hunk 5]
```diff
@@ -228,13 +233,42 @@ public void revokePermission(UUID bucketId, String userEmail) {
...
     currentUserHasPermissionForBucket(bucketId, AccessType.OWNER);
 
-    UserEntity user = userRepository.findUserByEmail(userEmail).orElseThrow(UserNotFoundException::new);
+    UserEntity user =
+        userRepository.findUserByEmail(userEmail).orElseThrow(UserNotFoundException::new);
+    checkIfTargetUserIsTheCurrentUser(user);
+    UserObjectAccessEntity permission =
+        permissionRepository
+            .findByObjectIdAndGrantedTo(bucketId, user)
+            .orElseThrow(BucketNotFoundException::new);
```
Comment (ivanfhtec) [inline backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java:242]: should be PermissionDoesNotExist

File: backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java [hunk 6]
```diff
@@ -215,8 +218,10 @@ public PermissionResponse editPermission(
       PermissionEditRequest permissionEditRequest, UUID bucketId) {
     checkIfBucketIsDeleted(bucketId);
     currentUserHasPermissionForBucket(bucketId, AccessType.EDIT);
-    UserObjectAccessEntity userObjectAccessEntity = permissionRepository.findByObjectId(bucketId);
+    UserObjectAccessEntity userObjectAccessEntity =
+        permissionRepository.findByObjectId(bucketId).orElseThrow(BucketNotFoundException::new);
```
Comment (ivanfhtec) [inline backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java:222]: should be PermissionDoesNotExist

File: backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java [hunk 7]
```diff
@@ -215,8 +218,10 @@ public PermissionResponse editPermission(
       PermissionEditRequest permissionEditRequest, UUID bucketId) {
     checkIfBucketIsDeleted(bucketId);
     currentUserHasPermissionForBucket(bucketId, AccessType.EDIT);
```
Comment (ivanfhtec) [inline backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java:220]: OWNER?

**Review by MetoMladenovski — APPROVED** [review]

File: backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java [hunk 8]
```diff
@@ -215,8 +218,10 @@ public PermissionResponse editPermission(
...
-    UserObjectAccessEntity userObjectAccessEntity = permissionRepository.findByObjectId(bucketId);
+    UserObjectAccessEntity userObjectAccessEntity =
+        permissionRepository.findByObjectId(bucketId).orElseThrow(BucketNotFoundException::new);
     UserEntity targetUser = findTargetUserForEdit(permissionEditRequest);
+    checkIfTargetUserIsTheCurrentUser(targetUser);
```
Comment (ivanfhtec) [inline backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java:224]: this looks like the target should be the current user
how about `checkCurrentUserNotTargetUser`?

File: backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java [hunk 9]
```diff
@@ -228,13 +233,42 @@ public void revokePermission(UUID bucketId, String userEmail) {
...
+    checkIfFileIsDeleted(fileMetadataEntity);
+    currentUserHasPermissionForFile(fileMetadataEntity, AccessType.OWNER);
+
+    UserEntity user = userRepository.findUserByEmail(email).orElseThrow(UserNotFoundException::new);
+    checkIfTargetUserIsTheCurrentUser(user);
     UserObjectAccessEntity permission =
-        permissionRepository.findByObjectIdAndGrantedTo(bucketId, user);
+        permissionRepository
+            .findByObjectIdAndGrantedTo(fileId, user)
+            .orElseThrow(FileNotFoundException::new);
```
Comment (ivanfhtec) [inline backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java:260]: PermissionDoesNotExist

File: backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java [hunk 10]
```diff
@@ -228,13 +233,42 @@ public void revokePermission(UUID bucketId, String userEmail) {
...
     checkIfPermissionExist(permission);
```
Comment (ivanfhtec) [inline backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java:261]: doesn't the line above check for this already?

File: backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java [hunk 11]
```diff
@@ -228,13 +233,42 @@ public void revokePermission(UUID bucketId, String userEmail) {
...
+    checkIfPermissionExist(permission);
+    permissionRepository.delete(permission);
+  }
+
+  @Override
+  public void revokeFilePermission(UUID fileId, String email) {
+    FileMetadataEntity fileMetadataEntity =
+        fileRepository.findById(fileId).orElseThrow(FileNotFoundException::new);
+    checkIfBucketIsDeleted(fileMetadataEntity.getBucket().getId());
+    checkIfFileIsDeleted(fileMetadataEntity);
```
Comment (ivanfhtec) [inline backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java:252]: if the bucket is deleted, `checkIfFileIsDeleted` would return true
I don't think you need `checkIfBucketIsDeleted`

**Pull request merged**